mappings/bnfdmd/data_downloader.py

```diff
@@ -0,0 +1,129 @@
+            if match:
+                inner_text = a_tag.text
+                while not (
+                    valid_from := datepat.search(inner_text.replace("\xa0", " "))
+                ):
+                    # sometimes "valid from" inner text is across adjacent elements
+                    inner_text += a_tag.next_sibling()[0].text
+                valid_from = datetime.strptime(valid_from.group(), datefmt)
+                matches.append((match.group("date"), url, filename, valid_from))
+
+        if not matches:
+            raise RuntimeError(f"Found no URLs matching {filename_re} at {self.url}")
+
+        # Sort by release date and get the latest
+        matches.sort()
+
+        return matches
+
+    def get_latest_release(self):
+        return self.get_releases()[-1]
+
+    def get_release_metadata(self, release):
+        datestamp, url, filename, valid_from = release
+
+        return {
+            "release": datestamp,
+            "valid_from": valid_from,
+            "url": url,
+            "filename": filename,
+            "release_name": datestamp,
+        }
+
+    def get_latest_release_metadata(self):
+        latest = self.get_latest_release()
+        return self.get_release_metadata(latest)
+
+    def download_release(self, release_name, valid_from):
+        """
+        Download a release that matches the specified release_name and valid_from values
+        """
+        logger.info(
+            "Attempting to download release from NHS BSA",
+            release=release_name,
+            valid_from=valid_from,
+        )
+
+        releases = self.get_releases()
+        release_metadata = (self.get_release_metadata(release) for release in releases)
+
+        match_found = False
+        for metadata in release_metadata:
+            if (
+                metadata.get("release_name") == release_name
+                and metadata.get("valid_from") == valid_from
+            ):
+                match_found = True
+                break
+
+        if not match_found:
+            raise ValueError(
+                f"No matching release found for release {release_name}, valid from {valid_from}"
+            )
+
+        local_download_filepath = Path(self.release_dir) / metadata["filename"]
+        self.get_file(metadata["url"], local_download_filepath)
+
+        return local_download_filepath
+
+    def get_file(self, url, filepath):
+        logger.info("Starting download", download_filepath=filepath)
+        with requests.get(url, stream=True) as resp:
+            resp.raise_for_status()
+            with open(filepath, "wb") as f:
+                for chunk in resp.iter_content(chunk_size=8192):
+                    f.write(chunk)
+        logger.info("Download complete", download_filepath=filepath)
+
+    def download_latest_release(self):
+        metadata = self.get_latest_release_metadata()
+
+        if glob.glob(os.path.join(self.release_dir, metadata["filename"])):
+            return
```

Comment:
question: Might it be possible that this gets in a broken state by, say:

* the file downloaded is broken (either the request is not completed for whatever reason, or the server gives an empty output)
* the file persists on disk
* and then further attempts to update for that release fail because the broken file is still there?

Fixing it would require someone to intervene, or waiting for the next release filename.

(For comparison, I am not completely sure how atomic the coding system imports are, and therefore how robust they are.)

Reply:
this was taken from the OP management command.

I think "manual intervention if it breaks" is fine in this case (deleting the broken download).

The alternative is to have the file be re-imported every time this command runs, regardless of whether there's a new update or not.

Since we don't (yet) record the version of the mapping file that's been loaded the only information we have as to the version is the filename itself.

Something to revisit if/when we start versioning the mappings